### PR TITLE
Add `Expr.valueOrAbort` and `reflect.report.errorAndAbort`

### PR DESCRIPTION
**Pull request by nicolasstucki**
Provides a homogeneous and unambiguous concept for stopping the expansion of a macro.

The advantages of this naming are

* Consistent name suffixes `xyzAbort`
* `report.e` will show auto-completion for all variants of `error` and `errorAndAbort`
* `Abort` cannot be confused with other kinds of error handling in this API

This is an alternative to #12022

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@liufengyun we also need to take into account the discussion in #12022

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -44,22 +44,22 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
      *  Otherwise returns the value.
      */
-    def valueOrError(using FromExpr[T]): T =
-      val fromExpr = summon[FromExpr[T]]
-      def reportError =
-        val msg = s"Expected a known value. \n\nThe value of: ${self.show}\ncould not be extracted using $fromExpr"
-        reflect.report.throwError(msg, self)
-      given Quotes = Quotes.this
-      fromExpr.unapply(self).getOrElse(reportError)
+    @deprecated("Use valueOrThrow", "3.0.0-RC3")
```
Comment (liufengyun) [inline library/src/scala/quoted/Quotes.scala:54]:
```suggestion
    @deprecated("Use valueOrAbort", "3.0.0-RC3")
```

**Review by liufengyun — APPROVED** [review]
LGTM

**Review by smarter — CHANGES_REQUESTED** [review]

File: library/src/scala/quoted/Quotes.scala [hunk 2]
```diff
@@ -44,22 +44,22 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+    def valueOrError(using FromExpr[T]): T = valueOrAbort
+
+    /** Return the value of this expression.
+     *
+     *  Emits an error and aborts if the expression does not represent a value or possibly contains side effects.
+     *  Otherwise returns the value.
+     */
+    def valueOrAbort(using FromExpr[T]): T
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:62]: I forgot to update that one to `3.1.0`. Done.

[the remark this answers is not in the document]
Comment (smarter) [reply library/src/scala/quoted/Quotes.scala:62]: Ah, if this PR is not supposed to be merged until 3.1.0 then we should mark it as draft for now.

**Review by liufengyun — APPROVED** [review]
LGTM

**Review by abgruszecki — APPROVED** [review]
LGTM!

**Comment by nicolasstucki** [comment]
Now the added methods are marked as experimental and we do not need to wait until 3.1 to add them.

**Comment by nicolasstucki** [comment]
Will need rebase on #12371

File: library/src/scala/quoted/Quotes.scala [hunk 3]
```diff
@@ -45,22 +45,24 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+    // @deprecated("Use valueOrThrow", "3.1.0")
+    def valueOrError(using FromExpr[T]): T
+
+    /** Return the value of this expression.
+     *
+     *  Emits an error and aborts if the expression does not represent a value or possibly contains side effects.
+     *  Otherwise returns the value.
+     */
+    @experimental
+    def valueOrAbort(using FromExpr[T]): T
```
Comment (liufengyun) [inline library/src/scala/quoted/Quotes.scala:66]: I'd suggest the method to take a message for indicating the reason why it's aborted. It helps end-users in debugging and saves the efforts of macro-authors in writing error-handling code.

**Review by liufengyun — APPROVED** [review]
LGTM

File: library/src/scala/quoted/Quotes.scala (hunk 3 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:66]: The question then is if we should keep the version with the default error message. The idea if this signature was also to make the authors not need to write the error message.

Currently the can do 
```scala
x.value match
  case Some(n) => ...
  case None => reflect.report.error("......", x)
```
or just
```scala
if x.valueOrAbort == 0 then ...
else 
```

the proposed one would need to add the message
```scala
if x.valueOrAbort(".........") == 0 then
else ...
```

Comment (liufengyun) [reply library/src/scala/quoted/Quotes.scala:66]: Indeed, the version with the error message does not look better in code and can be a little annoying for prototyping. For macro authors, it's not clear that it's better.

Macros can be subtle to use,  more debuggability to end-users will make them more friendly for usage. However, if the gain for end-users is small, maybe it's not worth the complication.

> The question then is if we should keep the version with the default error message

We probably only want to keep one of the two.

Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:66]: We could add an `valueOrAbortWith("....")` later or users will just define their own in the meantime. We should probably wait and see what users define themselves to know what is preferable in paractice.

Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:66]: Meanwhile, they can just define something like this in their projects

```scala
extension [T: FromExpr](expr: Expr[T]) 
  def valueOrAbortWith(msg: String)(using Quotes) = 
    x.value match
      case Some(n) => n
      case None => reflect.report.errorAndAbort(msg, x)
```

```scala
Expr(1).valueOrAbortWith("expected a constant")
```

Comment (liufengyun) [reply library/src/scala/quoted/Quotes.scala:66]: I agree, it's better to wait and do it later.

**Comment by nicolasstucki** [comment]
@liufengyun could you review the changes. I had to move some implementations to make MiMa happy.

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**